Fixes #19440 - Integration test provisioning discovered hosts

bastilian opened the pull request:
This is a start to add integration tests and to have a base for scenarios of bugs.
The first tests cover the provisioning modal and ensure it passes on the parameters, including the hostgroup_id, as a prerequisite for setting inherited values like the domain[1]

The test data still needs to be rewritten to use the factories for hostgroup, location and organization.

I would like to add a few more tests for the overview page, before it is done.
Are there any other tests that might be good to have right away?

[1] http://projects.theforeman.org/issues/19445

lzap commented on test/integration/discovered_hosts_test.rb:
```diff
@@ -0,0 +1,113 @@
+  let(:discovered_hosts) { Host::Discovered.all }
+
+  setup do
+    discovered_host.save
```
I'd perhaps do `save!` so we are sure there is at least one host, I mean if this breaks, all the mass-action tests like reboot might just pass silently.

lzap commented on test/integration/discovered_hosts_test.rb:
```diff
@@ -0,0 +1,113 @@
+
+  describe 'can provision discovered_host' do
+    setup do
+      page.find_link('Provision').click
```
So this click on random Provision link or the first one? Wondering if there are multiple records what happens.

lzap commented on test/integration/discovered_hosts_test.rb:
```diff
@@ -0,0 +1,113 @@
+      let(:discovery_hostgroup) { Hostgroup.first }
+
+      test 'it passes it on' do
+        select_from('host_hostgroup_id', discovery_hostgroup.name)
```
Name vs id is intentional?

lzap commented on test/integration/discovered_hosts_test.rb:
```diff
@@ -0,0 +1,113 @@
+    end
+  end
+
+  describe 'can provision discovered_host' do
```
I'd add "using Create host link".

lzap reviewed (commented):
Good start, but you seem to check parameters only, what usually breaks is that the form won't accept the entry and errors out via either Rails errors or hard 500 error.

Would it be possible to simply submit the form and then check we landed on Host Detail page somehow?

bastilian commented:
@lzap I've addressed the comments you had.

I will add more tests to touch the parts that prepare the host form and test as you suggested that the host form is rendered.

lzap commented:
LGTM

bastilian commented:
@lzap I've added a test for the form itself and to verify the attributes are set from the hostgroup.

For some reason now the issue with the domain (and other attributes) not being set correctly in the Integration test, however when performing the same manually in the UI, the attributes **are** correctly set.

I've added a fix[1] for it, but i hope to find out why it is happening in the tests now and have a better one soon. 

[1] https://github.com/theforeman/foreman_discovery/pull/354/files#diff-0e08314173f533a2b6e73218bc5d4b10R202

ares commented:
It seems tests are broken - `NoMethodError: undefined method `notifications' for nil:NilClass...`, not sure if this is caused by the PR though

lzap commented:
I think I found the issue - if you look into `domain` factory it always sets organization and location explicitly (Organization 1/Location 1) but you discover the host without any taxonomy. Depending of the context of the app, Domain will not likely present the value so you can't really choose it.

When you test this in UI hosts are always assigned an organization (there is a fallback mechanism - taxonomy of detected subnet is used, if not found then initial taxonomy is assigned). So you can't really ever discover a host without taxonomy.

Can you try again with setting taxonomy explicitly? Then remove the domain workaround you implemented. Maybe we should add a DiscoveredHost validator that makes sure taxonomy is always set, this can break some old tests we have tho.

lzap commented:
Rebased and trying it locally but getting this:

    Can't find entry point 'vendor' in webpack manifest

Any clue?

bastilian commented:
@lzap I encounter(ed) the same issue. The issue in my case was caused by webpack, that returened "Invalid Host header" for assets. If that is in your case too, a remedy is to disable the host check in webpack with adding `disableHostCheck: true` to [webpack.config.js](https://github.com/theforeman/foreman/blob/develop/config/webpack.config.js#L150)

bastilian commented:
@lzap The integration tests now have Location and Organization set and the domain check passes now without my attempted fix.

lzap commented on db/migrate/20150714144500_review_discovery_permissions.rb:
```diff
@@ -1,12 +1,14 @@
+        perms << "create_discovery_rules" if Permission.find_by_name("new_discovery_rules")
+        perms << "destroy_discovery_rules" if Permission.find_by_name("delete_discovery_rules")
+        mgr.add_permissions!(perms)
+      end
```
Please rebase and this can go I believe.

lzap commented on test/integration/discovered_hosts_test.rb:
```diff
@@ -0,0 +1,158 @@
+
+  setup do
+    set_default_settings
+    @discovered_host = FactoryGirl.create(:discovered_host, :with_facts)
```
Our API explicitly creates taxonomy for discovered hosts, I was wondering if you could move the taxonomy let blocks here and set it explicitly here as well.

lzap commented:
Allright core fixed let's re [test]

lzap commented:
Mind rebase? :-) Getting errors tho:

```
  1) Error:
DiscoveredHostsTest::edit form#test_0001_it is a host form:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

  2) Error:
NewHostNotificationTest#test_0002_multiple discovered hosts should generate only one notification:
NoMethodError: undefined method `notifications' for nil:NilClass
    /home/lzap/work/foreman_discovery/test/unit/ui_notifications/new_host_test.rb:13:in `block in <class:NewHostNotificationTest>'

  3) Failure:
NewHostNotificationTest#test_0001_new discovered host should generate a notification [/home/lzap/work/foreman_discovery/test/unit/ui_notifications/new_host_test.rb:5]:
Expected nil to be truthy.

  4) Error:
DiscoveredHostsTest::edit form::with a hostgroup passed#test_0001_sets inherited attributes:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

  5) Error:
DiscoveredHostsTest::Delete hosts#test_0001_it removes all hosts:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

  6) Error:
DiscoveredHostsTest::Autoprovision all#test_0001_converts all discovered to managed hosts:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

  7) Error:
DiscoveredHostsTest::using Create Host link::with a Location selected#test_0001_it passes it on:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

  8) Error:
DiscoveredHostsTest::using Create Host link::with a Hostgroup selected#test_0001_it passes it on:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

  9) Error:
DiscoveredHostsTest::using Create Host link::with a Organization selected#test_0001_it passes it on:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

 10) Error:
DiscoveredHostsTest::using Create Host link#test_0001_and forwards to editing it:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

 11) Error:
DestroyHostNotificationTest#test_0001_destorying discovered host should remove notification:
NoMethodError: undefined method `notifications' for nil:NilClass
    /home/lzap/work/foreman_discovery/test/unit/ui_notifications/destroy_host_test.rb:6:in `block in <class:DestroyHostNotificationTest>'
    /home/lzap/work/foreman_discovery/test/unit/ui_notifications/destroy_host_test.rb:6:in `block in <class:DestroyHostNotificationTest>'

 12) Error:
DestroyHostNotificationTest#test_0003_discovered host type change should update notifications:
NoMethodError: undefined method `notifications' for nil:NilClass
    /home/lzap/work/foreman_discovery/test/unit/ui_notifications/destroy_host_test.rb:24:in `block in <class:DestroyHostNotificationTest>'

 13) Error:
DestroyHostNotificationTest#test_0002_do not remove notification if others discovered hosts exists:
NoMethodError: undefined method `notifications' for nil:NilClass
    /home/lzap/work/foreman_discovery/test/unit/ui_notifications/destroy_host_test.rb:13:in `block in <class:DestroyHostNotificationTest>'

 14) Error:
DestroyHostNotificationTest#test_0004_type change should not removing notifications for discovered hosts:
NoMethodError: undefined method `notifications' for nil:NilClass
    /home/lzap/work/foreman_discovery/test/unit/ui_notifications/destroy_host_test.rb:33:in `block in <class:DestroyHostNotificationTest>'

 15) Error:
DiscoveredHostsTest::Reboot all#test_0001_triggers reboot on all discovered_hosts:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:21648
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:30
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:36587/assets/application-e1021443bec0dd4e5208fa108d2b962dcd1afe9b19c2c2128ea4b76a8199e07f.js:35 in onLoad
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:36587/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'

```

lzap commented:
I rebased ontop of develop which is finally green (again yeah!) but still getting these JS errors.

bastilian commented:
@lzap The JS errors are gone. There are (now) still tests failing:

```
  1) Error:
DiscoveredHostsTest::edit form#test_0001_it is a host form:
Capybara::Poltergeist::StatusFailError: Request to 'http://127.0.0.1:33743/discovered_hosts' failed to reach server, check DNS and/or server status - Timed out with no open resource requests
    test/test_helper.rb:35:in `before_setup'

  2) Error:
DiscoveredHostsControllerTest#test_edit_form_quick_submit:
ActiveRecord::RecordNotFound: Couldn't find Host::Managed with 'id'=33 [WHERE "hosts"."type" IN ('Host::Managed')]
    app/models/host.rb:15:in `method_missing'
    test/active_support_test_case_helper.rb:113:in `disable_taxonomies'

  3) Error:
NewHostNotificationTest#test_0002_multiple discovered hosts should generate only one notification:
NoMethodError: undefined method `notifications' for nil:NilClass
    

  4) Failure:
NewHostNotificationTest#test_0001_new discovered host should generate a notification [/var/lib/workspace/workspace/test_plugin_pull_request/database/postgresql/label/fast/ruby/2.2/plugin/test/unit/ui_notifications/new_host_test.rb:5]:
Expected nil to be truthy.
```

1. is an issue which happens in core tests as well without a fix for it yet. The other test seem to fail, as they miss seed data. 

These failing tests should probably be fixed in another PR, or?

tbrisker commented on test/integration/discovered_hosts_test.rb:
```diff
@@ -0,0 +1,148 @@
+require 'integration_test_helper'
+
+class DiscoveredHostsTest < IntegrationTestWithJavascript
+  let(:discovered_host) { FactoryGirl.create(:discovered_host, :with_facts) }
```
This needs to be changed to FactoryBot

tbrisker commented on test/integration/discovered_hosts_test.rb:
```diff
@@ -0,0 +1,148 @@
+      let(:hostgroup_environment) { FactoryGirl.create(:environment) }
+      let(:hostgroup_domain) { FactoryGirl.create(:domain) }
+      let(:hostgroup) do
+        FactoryGirl.create(:hostgroup, :with_os,
```
these too

lzap commented:
Yeah thanks, finally getting there! :)

bastilian commented:
@tbrisker Change to FactoryBot is done.

lzap commented:
Aaaand core is broken again :-)

lzap commented:
Ok it's our test what is failing, why blueprint is now `nil` ?

NoMethodError: undefined method `notifications' for nil:NilClass
     (NoMethodError)
/var/lib/workspace/workspace/test_plugin_pull_request@2/database/mysql/label/fast/ruby/2.2/plugin/test/unit/ui_notifications/destroy_host_test.rb:6

lzap commented:
Ok tests are failing locally for the same reason, please take a look.

ares commented:
It seems there are still some tests failing :-(

lzap commented:
I am getting these locally:

```
Rails logs for test_0001_it passes it on FAILURE:
E

Error:
DiscoveredHostsTest::using Create Host link::with a Location selected#test_0001_it passes it on:
Capybara::Poltergeist::JavascriptError: One or more errors were raised in the Javascript code on the page. If you don't care about these errors, you can ignore them by setting js_errors: false in your Poltergeist configuration (see documentation for details).

TypeError: undefined is not an object (evaluating 'tfm.nav.init')
TypeError: undefined is not an object (evaluating 'tfm.nav.init')
    at http://127.0.0.1:39927/assets/application-daecd4ed7147767e48554589a4bced4d637accc3beef1df5d169160640c1d097.js:21648
    at http://127.0.0.1:39927/assets/application-daecd4ed7147767e48554589a4bced4d637accc3beef1df5d169160640c1d097.js:30
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29 in dispatch
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29 in handle
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:19 in each
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29 in trigger
    at http://127.0.0.1:39927/assets/application-daecd4ed7147767e48554589a4bced4d637accc3beef1df5d169160640c1d097.js:35 in onLoad
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29 in l
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29 in fireWith
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:29 in ready
    at http://127.0.0.1:39927/webpack/bundle-cf97ad12463c31ce45f1.js:19 in u
    /home/lzap/work/foreman_discovery/test/integration/discovered_hosts_test.rb:11:in `block in <class:DiscoveredHostsTest>'
    test/test_helper.rb:35:in `before_setup'
```

tbrisker commented:
@lzap - that looks like you need to update your foreman and recompile webpack (`rake webpack:compile`)

bastilian commented:
There was one test still failing on CI, which failed locally as well.
It seemed to have been test failing by chance when taxonomies are disabled and the discovered host is in a hostgroup with a subnet not that invalidated the IP provided in the facts for importing. 

All other test passed locally and should now on CI pass as well.

ares commented:
[test] to resurrect from dns issues

bastilian committed:
Fixes #19440 - Integration test provisioning discovered hosts

bastilian commented:
@ares :green_heart: once again! \o/

ares commented:
Awesome, @lzap can we merge while it's green? :-)

ares commented:
@tbrisker could you please update your review? :-)

tbrisker commented:
@ares If any of my reviews is out of date and already addressed feel free to dismiss it :)

ares commented:
yeah, I don't have permissions to do it here :-)

lzap commented:
Ok looks good to me, thanks @bastilian for such an effort.
